File: src/Test/ArgumentMatchingWithSubscriberConverter.cs
using SecretNest.MessageBus.Options;
using System;

namespace Test
{
    [TestClass]
    public class ArgumentMatchingWithSubscriberConverter
    {
        [TestMethod]
        public void TestMethod()
        {
            using var bus = new MessageBus();
            var subscriber1000 = bus.RegisterSubscriber<string, string>("ArgumentMatching", Subscriber1000,
                new MessageBusSubscriberOptions<string, string>(
                    sequence: 0,
                    resultCheckingCallback: i => i != null,
                    argumentConvertingCallback: i => i?.ToString(),
                    returnValueConvertingCallback: s => int.Parse(s!)));
            var subscriber100 = bus.RegisterSubscriber<int, int>("ArgumentMatching", Subscriber100,
                new MessageBusSubscriberOptions<int, int>(
                    sequence: 1,
                    resultCheckingCallback: i => i > 0));
            
            var publisher = bus.RegisterPublisher<int, int>("ArgumentMatching", new MessageBusPublisherOptions<int, int>(defaultReturnValue: 5));

            var result1000 = publisher.Executor.Execute(1000);
            var result100 = publisher.Executor.Execute(300);
            var result5 = publisher.Executor.Execute(80);

            bus.UnregisterPublisher(publisher);
            bus.UnregisterSubscriber(subscriber1000);
            bus.UnregisterSubscriber(subscriber100);

            Assert.AreEqual(1000, result1000);
            Assert.AreEqual(100, result100);
            Assert.AreEqual(5, result5);
        }

        public string? Subscriber1000(string? argument)
        {
            if (argument == "1000")
                return "1000";
            else
                return null;
        }

        public int Subscriber100(int argument)
        {
            if (argument > 100)
                return 100;
            else
                return 0;
        }
    }
}

File: src/TestVB/ArgumentMatchingWithSubscriberConverter.vb
Imports Microsoft.VisualStudio.TestTools.UnitTesting
Imports SecretNest.MessageBus
Imports SecretNest.MessageBus.Options

<TestClass>
Public Class ArgumentMatchingWithSubscriberConverter
    <TestMethod>
    Public Sub TestMethod()
        Using bus As New MessageBus
            Dim subscriber1000Ticket = bus.RegisterSubscriber(Of String, String)("ArgumentMatching", AddressOf Subscriber1000, New MessageBusSubscriberOptions(Of String, String)(sequence:=0, resultCheckingCallback:=Function(i) i IsNot Nothing, argumentConvertingCallback:=Function(i) i?.ToString(), returnValueConvertingCallback:=Function(s) Integer.Parse(s)))
            Dim subscriber100Ticket = bus.RegisterSubscriber(Of Integer, Integer)("ArgumentMatching", AddressOf Subscriber100, New MessageBusSubscriberOptions(Of Integer, Integer)(sequence:=1, resultCheckingCallback:=Function(i) i > 0))

            Dim publisherTicket = bus.RegisterPublisher(Of Integer, Integer)("ArgumentMatching", New MessageBusPublisherOptions(Of Integer, Integer)(defaultReturnValue:=5))

            Dim result1000 = publisherTicket.Executor.Execute(1000)
            Dim result100 = publisherTicket.Executor.Execute(300)
            Dim result5 = publisherTicket.Executor.Execute(80)

            bus.UnregisterPublisher(publisherTicket)
            bus.UnregisterSubscriber(subscriber1000Ticket)
            bus.UnregisterSubscriber(subscriber100Ticket)

            Assert.AreEqual(1000, result1000)
            Assert.AreEqual(100, result100)
            Assert.AreEqual(5, result5)
        End Using
    End Sub

    Public Shared Function Subscriber1000(argument As String) As String
        If argument = "1000" Then
            Return "1000"
        Else
            Return Nothing
        End If
    End Function

    Public Shared Function Subscriber100(argument As Integer) As Integer
        If argument > 100 Then
            Return 100
        Else
            Return 0
        End If
    End Function
End Class
